tools/src/abstract_server/server_interface.rs

```diff
@@ -0,0 +1,96 @@
+use async_trait::async_trait;
+use serde_json::Value;
+
+pub type Result<T> = std::result::Result<T, ServerError>;
+
+/// Express whether the error seems to be happening in the server or the data.
+#[derive(Debug)]
+pub enum ErrorLayer {
+    /// The request itself has issues like a bad URL.
+    BadInput,
+    /// The error seems to involve server logic, so it may or may not be an issue
+    /// with the underlying data.
+    ServerLayer,
+    /// The error seems to be related to the indexed data in question rather
+    /// than the server, like the data was not indexed.
+    DataLayer,
+    /// We're not sure if it was a server issue or a data issue.
+    UnknownLayer,
+}
+
+/// ServerError payload to provide details about what went wrong for
+/// investigation purposes.  In the future, this could wrap the
+/// underlying errors we've seen.
+#[derive(Debug)]
+pub struct ErrorDetails {
```

Comment:
I really like the way this error stuff is structured. And really nice documentation, thanks!

Reply:
Thanks!